src/storage/src/hummock/sstable/restricted_builder.rs

```diff
@@ -0,0 +1,208 @@
+
+    writer: Arc<SstWriter>,
+
+    limiter: Arc<ResourceLimiter>,
+
+    vnode_bitmaps: Vec<Vec<VNodeBitmap>>,
+
+    resource_size: u64,
+}
+
+impl<B, F> RestrictedBuilder<B>
+where
+    B: Clone + Fn() -> F,
+    F: Future<Output = HummockResult<(HummockSSTableId, SSTableBuilder)>>,
+{
+    pub fn new(
+        get_id_and_builder: B,
+        writer: Arc<SstWriter>,
+        limiter: Arc<ResourceLimiter>,
+        resource_size: u64,
+    ) -> Self {
+        Self {
+            get_id_and_builder,
+            current_builder: None,
+            write_results: Default::default(),
+            writer,
+            limiter,
+            vnode_bitmaps: Default::default(),
+            resource_size,
+        }
+    }
+
+    pub fn len(&self) -> usize {
+        let mut len = self.write_results.len();
+        if self.current_builder.is_some() {
+            len += 1;
+        }
+        len
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.current_builder.is_none() && self.write_results.is_empty()
+    }
+
+    pub async fn add_full_key(
+        &mut self,
+        full_key: FullKey<&[u8]>,
+        value: HummockValue<&[u8]>,
+        allow_split: bool,
+    ) -> HummockResult<()> {
+        let new_builder_required = match self.current_builder.as_ref() {
+            None => true,
+            Some(b) => allow_split && (b.builder.reach_capacity() || b.sealed),
+        };
+
+        if new_builder_required {
+            self.finish_current_builder();
+            let (id, builder) = (self.get_id_and_builder)().await?;
+            self.current_builder = Some(SSTableBuilderWrapper {
+                id,
+                builder,
+                sealed: false,
+            });
+            let (rx, tx) = tokio::sync::oneshot::channel();
+            self.limiter.request_resource(self.resource_size, rx);
+            if let Err(e) = tx.await {
+                return Err(HummockError::other(format!(
+                    "Cannot request resource.\n{:#?}",
+                    e
+                )));
+            }
+        }
+
+        let builder = &mut self.current_builder.as_mut().unwrap().builder;
+        builder.add(full_key.into_inner(), value);
+        Ok(())
+    }
+
+    pub fn seal_current(&mut self) {
+        if let Some(b) = &mut self.current_builder {
+            b.sealed = true;
+        }
+    }
+
+    fn finish_current_builder(&mut self) {
+        if let Some(current_builder) = self.current_builder.take() {
+            let id = current_builder.id;
+            let (data, meta, vnode_bitmaps) = current_builder.builder.finish();
+            let (tx, rx) = tokio::sync::oneshot::channel::<HummockResult<Sstable>>();
+            self.writer.write(Sstable { id, meta }, data, tx);
+            self.vnode_bitmaps.push(vnode_bitmaps);
+            let limiter = self.limiter.clone();
+            let resource_size = self.resource_size;
+            self.write_results.push(tokio::spawn(async move {
```

Comment:
May use `then` instead execute `spawn`  reduce the cost of tokio schedule?

Reply:
I think spawn is needed here. Because I want to get the write_result sooner in order to free the resource, meanwhile I don't want to block the caller of `finish_current_builder`.